app/external/pubsub.py

```
@@ -1,28 +1,62 @@
 from google.cloud import pubsub_v1
-import os
-from typing import Dict, Optional
+import os, contextlib
+from typing import Dict, List, Optional
 
-_client: Optional[pubsub_v1.PublisherClient] = None
+_publisher_client: Optional[pubsub_v1.PublisherClient] = None
+_subscriber_client: Optional[pubsub_v1.SubscriberClient] = None
 
 
-def _get_client() -> pubsub_v1.PublisherClient:
-    global _client
-    if _client is None:
-        _client = pubsub_v1.PublisherClient()
-    return _client
+def _get_publisher_client() -> pubsub_v1.PublisherClient:
+    global _publisher_client
+    if _publisher_client is None:
+        _publisher_client = pubsub_v1.PublisherClient()
+    return _publisher_client
+
+
+def create_topic_and_sub() -> None:
+    pubclient = _get_publisher_client()
+    subclient = _get_subscriber_client()
+    topic_path = pubclient.topic_path(os.environ.get("PUBSUB_PROJECT"), os.environ.get("PUBSUB_TOPIC"))
+    sub_path = subclient.subscription_path(os.environ.get("PUBSUB_PROJECT"), os.environ.get("PUBSUB_SUBSCRIPTION"))
+
+    with contextlib.suppress(Exception): # it's fine if the topic already exists.
+        pubclient.create_topic(topic_path)
+
+    with contextlib.suppress(Exception): # it's fine if the subscription already exists.
```

Would this suppress exceptions that aren't the topic/sub already existing?

---

yes, but they'll be limited entirely to google-response-exceptions that we couldn't do anything about. google's pubsub library retries failures appropriately so we don't have to worry about handling transient things.